File: windows/BlueLink.App/Updates/ReleaseVersion.cs
using System.Text.RegularExpressions;

namespace BlueLink.Updates;

/// <summary>The release tag identifies preview builds independently of the numeric assembly version.</summary>
public sealed record ReleaseVersion(Version Version, int? Preview) : IComparable<ReleaseVersion>
{
    public string Tag => "v" + Version + (Preview is { } number ? "-preview." + number : "");
    public int CompareTo(ReleaseVersion? other)
    {
        if (other is null) return 1;
        var comparison = Version.CompareTo(other.Version);
        if (comparison != 0) return comparison;
        if (Preview is null) return other.Preview is null ? 0 : 1;
        return other.Preview is null ? -1 : Preview.Value.CompareTo(other.Preview.Value);
    }
    public static ReleaseVersion? Parse(string? tag)
    {
        var match = Regex.Match(tag ?? "", @"^v?(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(?:-preview\.([1-9][0-9]*))?$");
        if (!match.Success || !int.TryParse(match.Groups[1].Value, out var major) ||
            !int.TryParse(match.Groups[2].Value, out var minor) || !int.TryParse(match.Groups[3].Value, out var patch) ||
            (match.Groups[4].Success && !int.TryParse(match.Groups[4].Value, out _))) return null;
        return new(new(major, minor, patch), match.Groups[4].Success ? int.Parse(match.Groups[4].Value) : null);
    }
}

File: android/app/src/main/java/com/bluelink/android/updates/ReleaseVersion.kt
package com.bluelink.android.updates

internal data class ReleaseVersion(val major: Int, val minor: Int, val patch: Int, val preview: Int?) : Comparable<ReleaseVersion> {
    val base: String get() = "$major.$minor.$patch"
    val tag: String get() = "v" + base + (preview?.let { "-preview.$it" } ?: "")
    override fun compareTo(other: ReleaseVersion): Int {
        compareValuesBy(this, other, { it.major }, { it.minor }, { it.patch }).let { if (it != 0) return it }
        if (preview == null) return if (other.preview == null) 0 else 1
        return other.preview?.let { preview.compareTo(it) } ?: -1
    }
    companion object {
        fun parse(tag: String): ReleaseVersion? {
            val match = Regex("""^v?(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(?:-preview\.([1-9][0-9]*))?$""").matchEntire(tag) ?: return null
            val parts = (1..3).map { match.groupValues[it].toIntOrNull() ?: return null }
            val preview = match.groupValues[4].takeIf { it.isNotEmpty() }?.let { it.toIntOrNull() ?: return null }
            return ReleaseVersion(parts[0], parts[1], parts[2], preview)
        }
    }
}
